### DISCUSSION TITLE
Thread safe property access

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/runtime#46318] [Q&A] [answered]
Is there any guarantee that property access always reads the field from memory ?

E.g. this code from .NET.
```cs
internal bool Disposed => _disposed != 0;
```

My assumption that JIT can optimize the property access into a variable access or into register and then the processor might cache its value.

**Comment by PathogenDavid** [comment]
I don't believe properties do anything special here.

If you need this guarantee you should mark `_disposed` as [`volatile`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/volatile) to indicate to the JIT that it shouldn't make any optimizations that assume single-threaded access.

**Comment by acaly** [comment]
You might be aware of this, but I still want to point out that `volatile` by itself doesn't guarantee thread-safety. You'll have to use either a lock or an atomic RMW operation (c# has an `Interlocked` class). Those are obviously more than an auto-property should handle.

**Reply by svick** [reply]
@acaly It does in C# (which is different from how `volatile` behaves in C and C++).

Here is [the spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes?WT.mc_id=DOP-MVP-5002735#volatile-fields):

> *  A read of a volatile field is called a ***volatile read***. A volatile read has "acquire semantics"; that is, it is guaranteed to occur prior to any references to memory that occur after it in the instruction sequence.
> *  A write of a volatile field is called a ***volatile write***. A volatile write has "release semantics"; that is, it is guaranteed to happen after any memory references prior to the write instruction in the instruction sequence.
>
> These restrictions ensure that all threads will observe volatile writes performed by any other thread in the order in which they were performed. A conforming implementation is not required to provide a single total ordering of volatile writes as seen from all threads of execution.

**Reply by acaly** [reply]
I know it's different from volatile in C++, but no, volatile by itself usually isn't thread safe. 

> A read of a volatile field ... it is guaranteed to occur prior to any references to memory that occur after it **in the instruction sequence**
> A write of a volatile field ... it is guaranteed to happen after any memory references prior to the write instruction **in the instruction sequence**

These are basically talking about ordering within one single thread. When there are multiple threads, all reads might happen earlier and all writes might happen later, and the thread might be preempted by OS or runtime GC. 

Note that the simplest spin lock in C++ uses not only the `std::atomic_flag` (a volatile bool), but also the `atomic_flag::test_and_set` (an atomic RMW operation). C++ has `std::atomic<T>::load` and `std::atomic<T>::store`, which accept `std::memory_order`, including the acquire and release semantics (with the same power of C# volatiles), but still, `test_and_set` is necessary to implement a lock. (The underlying principle is the [consensus number](https://en.wikipedia.org/wiki/Consensus_(computer_science)#Shared-memory_consensus_protocols)).

I can imagine 2 uses of the `_disposed` flag:

1. Check the flag and dispose the object if it's not set yet.
2. Check the flag and access its content if it's not set.

In both cases, the volatile read cannot protect the object from being disposed by other threads.

**Reply by acaly** [reply]
You can confirm that in C#, a volatile read on a `int` isn't really different from a normal read from [generated code](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgMwAEu5AwuQN7HlPkAOCAlgG4TzmcD2AGx7tBccuwB2McgH0AZsIDmeANyNmZcsH5DyAWTwAKAJTkAvAD45iiCvIBCc+QAM6os1YduvKTIXKOO6eWjp6+jimFtYBdjiOzm7EAL5AA).

**Reply by gfoidl** [reply]
> You can confirm that in C#, a volatile read on a `int` isn't really different from a normal read from 

Sharplab shows the codgen for x86/x64, and they have a strong memory model.
On ARM -- which has a weak memory model -- the **JIT** (not C# compiler Roslyn) takes this into account and generates different code from normal reads/writes. It's only on strong memory models, where the "volatile evaporates".

**Reply by acaly** [reply]
> they have a strong memory model

I understand that. Still, this stronger memory model isn't enough for thread safety. Why do you think we still have RMW and locks on x86/x64 platforms?

**Reply by gfoidl** [reply]
To note: my reply was just for the quoted text (which is correct, but not in general, hence the addendum for it).

Thread safety / atomicity is one "thing", the other one is visibility. 
For the latter memory models come into play.
For the first only read / writes with word-size are atomic, all other read / writes need to be done with special treatment, like Interlocked (asm on x86: `lock cmpxchg` so at the cpu-level + full memory barrier) or by exclusive access to the shared ressource like monitors, etc. where the runtime also guarantees a memory barrier, so that the change gets "visible".

So `volatile` is for visibility, and has per se nothing to do with atomicity.
(That current .NET uses Interlocked for some implementations of Volatile is an implemenation detail and hence atomicity is a side-effect).

**Comment by danmoseley** [comment, marked as answer]
Also answered by @gfoidl and @PathogenDavid
